quotest/quotest.cpp

```diff
@@ -871,6 +876,35 @@ TEST_IMPL(visitResources)
     FINISH_TEST(true);
 }
 
+TEST_IMPL(thread)
+{
+    auto rootTxnId = targetRoom->postPlainText("Threadroot"_L1);
+    connect(targetRoom, &Room::pendingEventAboutToMerge, this, [this, thisTest, rootTxnId](Quotient::RoomEvent* rootEvt) {
+        if (rootEvt->transactionId() == rootTxnId) {
+            const auto relation = EventRelation::replyInThread(rootEvt->id(), true, rootEvt->id());
+            auto replyTxnId = targetRoom->post<RoomMessageEvent>(u"Thread reply 1"_s, RoomMessageEvent::MsgType::Text, nullptr, relation)->transactionId();
+            connect(targetRoom, &Room::pendingEventAboutToMerge, this, [this, thisTest, replyTxnId](Quotient::RoomEvent* replyEvt) {
+                if (replyEvt->transactionId() == replyTxnId) {
+                    connect(targetRoom, &Room::pendingEventMerged, this, [this, thisTest, replyEvt]() {
+                        replyEvt->switchOnType(
+                            [&](const RoomMessageEvent& rmReplyEvt) {
+                                const auto thread = targetRoom->threads()[rmReplyEvt.threadRootEventId()];
+                                FINISH_TEST(thread.threadRootId == rmReplyEvt.threadRootEventId() &&
+                                            thread.latestEventId == rmReplyEvt.id() &&
+                                            thread.size == 2
+                                );
+                            },
+                            [this, thisTest](const RoomEvent&) { FAIL_TEST(); }
+                        );
+                    });
+                }
+            });
```

Comment:
It can be written simpler using the future provided by `PendingEventItem` these days (and I hope all `post*()` methods eventually converge on returning a `PendingEventItem` instead of transaction id because it makes further actions on the pending event so much easier):
```suggestion
            targetRoom
                ->post<RoomMessageEvent>(u"Thread reply 1"_s, RoomMessageEvent::MsgType::Text,
                                         nullptr, relation)
                .whenMerged()
                .then([this, thisTest](const RoomEvent& replyEvt) {
                    replyEvt.switchOnType(
                        [&](const RoomMessageEvent& rmReplyEvt) {
                            const auto thread =
                                targetRoom->threads()[rmReplyEvt.threadRootEventId()];
                            FINISH_TEST(thread.threadRootId == rmReplyEvt.threadRootEventId()
                                        && thread.latestEventId == rmReplyEvt.id()
                                        && thread.size == 2);
                        },
                        [this, thisTest](const RoomEvent&) { FAIL_TEST(); });
                });
```

Reply:
It still needs to connect to `pendingEventMerged` as `whenMerged` goes off before the event is in the timeline so the thread will not have been created but otherwise OK.

Reply:
Hmm, that wasn't the intention of `whenMerged()` :-\ I really had an idea of that `whenMerged()` only completes when the ultimate event is in the timeline. I'll go fix that separately.

Reply:
#844 should rectify that, then you can rebase on it and get rid of the extra `connect()` round.